### PR TITLE
kernel/binary_manager: Support binary loading from file system

### PR DESCRIPTION
**Pull request by jeongarmy**
kernel/binary_manager: Support binary loading from file system

### CODE REVIEW

File: framework/include/binary_manager/binary_manager.h [hunk 1]
```diff
@@ -103,6 +103,36 @@ int binary_manager_register_state_changed_callback(binmgr_statecb_t handler, voi
  */
 int binary_manager_unregister_state_changed_callback(void);
 
+/**
+ * @brief Create a binary file with version.
+ * @details @b #include <binary_manager/binary_manager.h>\n
+ *  It creates and opens a file of binary which has a binary name with version.
+ * @return A defined value of binmgr_response_result_type in <tinyara/binary_manager.h>
+ *         fd of a created file on success. On failure, negative value is returned.
+ * @since TizenRT v3.0 PRE
```
Comment (jeongchanKim) [inline framework/include/binary_manager/binary_manager.h]: I think this version should be changed to v3.1 PRE or something.

**Comment by seinfra** [comment]
Target : [6773709d614acbf395f37afe7e4dbdce2001d8c8] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [6773709d614acbf395f37afe7e4dbdce2001d8c8] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [e875e20283d573e5a2ce3be66cf59d3d13ebb1d0] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [e875e20283d573e5a2ce3be66cf59d3d13ebb1d0] - Code Rule Check (C++) OK.

File: framework/src/binary_manager/binary_manager_update.c [hunk 2]
```diff
@@ -125,3 +126,72 @@ int binary_manager_get_update_info_all(binary_update_info_list_t *binary_info_li
...
+		ret = BINMGR_OPERATION_FAIL;
+	}
+	return ret;
+}
+
+int binary_manager_write_binary(int fd, FAR const void *buf, size_t nbytes)
+{
+	int ret;
+
+	ret = read(fd, buf, nbytes);
```
Comment (sijupunnnoose) [inline framework/src/binary_manager/binary_manager_update.c]: Is it read or write?

File: framework/src/binary_manager/binary_manager_interface.c [hunk 3]
```diff
@@ -46,6 +46,15 @@ int binary_manager_set_request(binmgr_request_t *request_msg, int cmd, void *arg
 		}
 		snprintf(request_msg->data.bin_name, BIN_NAME_MAX, "%s", (char *)arg);
 		break;
+	case BINMGR_CREATE_BIN:
+		if (arg == NULL) {
```
Comment (sijupunnnoose) [inline framework/src/binary_manager/binary_manager_interface.c]: How about moving arg null check at beginning of the function instead?

File: framework/src/binary_manager/binary_manager_update.c [hunk 4]
```diff
@@ -125,3 +126,72 @@ int binary_manager_get_update_info_all(binary_update_info_list_t *binary_info_li
...
+
+int binary_manager_open_binary(char *binary_name, int version)
+{
+	int fd;
+	int ret;
+	binmgr_update_bin_t data;
+	binmgr_request_t request_msg;
+	binmgr_createbin_response_t response_msg;
+
+	if (binary_name == NULL || version == NULL) {
```
Comment (sijupunnnoose) [inline framework/src/binary_manager/binary_manager_update.c]: version is not a pointer, why comparing with NULL?

File: os/kernel/binary_manager/binary_manager_load.c [hunk 5]
```diff
@@ -137,117 +141,157 @@ static int binary_manager_read_header(int bin_idx, int part_idx, binary_header_t
...
+	load_attr.priority = header_data->bin_priority;
+	load_attr.offset = CHECKSUM_SIZE + header_data->header_size;
+	strncpy(load_attr.bin_ver, header_data->bin_ver, BIN_VER_MAX);
+#ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
+	binp = BIN_LOADINFO(bin_idx);
+	/* If current version is different from version to be loaded, Clean up backup data */
+	if (binp && strncmp(BIN_VER(bin_idx), header_data->bin_ver, strlen(BIN_VER(bin_idx)) + 1)) {
+		binp->reload = false;
+		binfmt_exit(binp);
+		binp = NULL;
```
Comment (jeongchanKim) [inline os/kernel/binary_manager/binary_manager_load.c]: If we cleanup the binp, we don't need to set binp->reload to false.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 6]
```diff
@@ -137,117 +141,157 @@ static int binary_manager_read_header(int bin_idx, int part_idx, binary_header_t
...
-		bmdbg("Invalid binary state %d\n", BIN_STATE(bin_idx));
-		return ERROR;
-	}
+	file_idx = 0;
 
-	/* Read header data of binary partitions */
-	for (part_idx = 0; part_idx < PARTS_PER_BIN; part_idx++) {
-		if (BIN_PARTNUM(bin_idx, part_idx) < 0) {
-			continue;
+	bmdbg("Open a directory, %s\n", BINARY_DIR_PATH);
```
Comment (jeongchanKim) [inline os/kernel/binary_manager/binary_manager_load.c]: I think it can be bmvdbg.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 7]
```diff
@@ -137,117 +141,157 @@ static int binary_manager_read_header(int bin_idx, int part_idx, binary_header_t
...
 /****************************************************************************
- * Name: binary_manager_load_binary
+ * Name: binary_manager_exec_binary
  *
  * Description:
- *	 This function loads binary with index in binary table.
+ *	 This function loads and executes binary.
  *
  ****************************************************************************/
+static int binary_manager_exec_binary(int bin_idx, char *path, binary_header_t *header_data)
```
Comment (jeongchanKim) [inline os/kernel/binary_manager/binary_manager_load.c]: I think this function is not only for execution. It includes load also.
Why did you change the function name?

File: framework/src/binary_manager/binary_manager_interface.c (hunk 3 above)
Comment (jeongarmy) [reply framework/src/binary_manager/binary_manager_interface.c]: In case of BINMGR_UNREGISTER_STATECB, it doesn't need arg value.

File: framework/src/binary_manager/binary_manager_update.c [hunk 8]
```diff
@@ -125,3 +126,72 @@ int binary_manager_get_update_info_all(binary_update_info_list_t *binary_info_li
...
+		bmdbg("Invalid parameter\n");
+		return BINMGR_INVALID_PARAM;
+	}
+
+	strncpy(data.bin_name, binary_name, BIN_NAME_MAX);
+	data.version = version;
+
+	ret = binary_manager_set_request(&request_msg, BINMGR_CREATE_BIN, &data);
+	if (ret < 0) {
+		return BINMGR_COMMUNICATION_FAIL;
```
[the remark this answers is not in the document]
Comment (jeongarmy) [reply framework/src/binary_manager/binary_manager_update.c]: binary_manager_set_request includes debug message for call errors

File: os/kernel/binary_manager/binary_manager_load.c (hunk 5 above)
Comment (jeongarmy) [reply os/kernel/binary_manager/binary_manager_load.c]: It is checked and memory is freed in binfmt_exit.

**Comment by seinfra** [comment]
Target : [742ee6e216e028fb36c58a9ee1ee613ac2ec201e] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [742ee6e216e028fb36c58a9ee1ee613ac2ec201e] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [64bc778df749acf4b33e9872ef72560734ab9f14] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [64bc778df749acf4b33e9872ef72560734ab9f14] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [31d98658f5601bf661f5aa27fbee0efe88933d3d] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [31d98658f5601bf661f5aa27fbee0efe88933d3d] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [12ed077ec073016d2c6fddbf69c15ed350e5d0cd] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [12ed077ec073016d2c6fddbf69c15ed350e5d0cd] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [69e63fa7ffaea27c591ae1e24871231421918df6] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [69e63fa7ffaea27c591ae1e24871231421918df6] - Code Rule Check OK.

File: framework/include/binary_manager/binary_manager.h [hunk 9]
```diff
@@ -99,10 +99,21 @@ int binary_manager_register_state_changed_callback(binmgr_statecb_t handler, voi
  *  It clears registered callback for the changes of binary state.
  * @return A defined value of binmgr_response_result_type in <tinyara/binary_manager.h>
  *         0 (BINMGR_OK) On success. On failure, negative value is returned.
- * @since TizenRT v3.0 PRE
+ * @since TizenRT v3.0
  */
 int binary_manager_unregister_state_changed_callback(void);
 
+/**
+ * @brief Create a binary file with version.
```
Comment (jeongchanKim) [inline framework/include/binary_manager/binary_manager.h]: I think this expression can be confused to App developer.

File: os/kernel/binary_manager/binary_manager_getinfo.c [hunk 10]
```diff
@@ -98,6 +102,64 @@ int binary_manager_get_index_with_name(char *bin_name)
...
+			if (!entryp) {
+				/* Finished with this directory */
+				break;
+			}
+			if (DIRENT_ISFILE(entryp->d_type) && !strncmp(entryp->d_name, bin_name, name_len) \
+				&& entryp->d_name[name_len] == '_' && strncmp(entryp->d_name, running_file, strlen(running_file))) {
+				snprintf(filepath, BIN_FILEPATH_MAX, "%s/%s", BINARY_DIR_PATH, entryp->d_name);
+				if (stat(filepath, &file_buf) == OK) {
+					bmvdbg("filepath %s size %d\n", filepath, file_buf.st_size);
+					size += file_buf.st_size;
```
Comment (jeongchanKim) [inline os/kernel/binary_manager/binary_manager_getinfo.c]: Please add comment why we need this.

**Comment by seinfra** [comment]
Target : [7fa7f57163d0288974ac1c663fcbdaf34acab04b] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [7fa7f57163d0288974ac1c663fcbdaf34acab04b] - Code Rule Check (C++) OK.

File: os/kernel/binary_manager/binary_manager_binfile.c [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+		goto send_result;
+	}
+
+	snprintf(filepath, BIN_FILEPATH_MAX, "%s/%s_%d", BINARY_DIR_PATH, bin_name, version);
+	fd = open(filepath, O_RDWR | O_CREAT, 0666);
+	if (fd > 0) {
+		bmvdbg("Created file '%s' for binary %s\n", filepath, bin_name);
+		strncpy(response_msg.binpath, filepath, BIN_FILEPATH_MAX);
+		response_msg.result = BINMGR_OK;
+		close(fd);
```
Comment (jeongchanKim) [inline os/kernel/binary_manager/binary_manager_binfile.c]: In this success case, it also returns BINMGR_OPERATION_FAIL. Is it correct?

**Comment by seinfra** [comment]
Target : [fa93022429f415543d4bae65d92fc64de94ad90b] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fa93022429f415543d4bae65d92fc64de94ad90b] - Code Rule Check OK.

File: os/kernel/binary_manager/binary_manager_getinfo.c [hunk 12]
```diff
@@ -163,7 +163,7 @@ void binary_manager_get_info_all(int requester_pid)
 
 	bin_count = binary_manager_get_binary_count();
 	if (bin_count > 0) {
-		for (bin_idx = 0; bin_idx < bin_count + 1; bin_idx++) {
+		for (bin_idx = 0; bin_idx < bin_count; bin_idx++) {
```
Comment (YiHyunJin) [inline os/kernel/binary_manager/binary_manager_getinfo.c]: Does not receive kernel binary information?

**Comment by seinfra** [comment]
Target : [77f967bb1a3373dc2b3cf44044ed8e06cf077d49] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [77f967bb1a3373dc2b3cf44044ed8e06cf077d49] - Code Rule Check OK.

File: os/kernel/binary_manager/binary_manager_getinfo.c [hunk 13]
```diff
@@ -98,6 +102,65 @@ int binary_manager_get_index_with_name(char *bin_name)
...
+	snprintf(running_file, BIN_FILENAME_MAX, "%s_%s", BIN_NAME(bin_idx), BIN_VER(bin_idx));
+	bin_name = BIN_NAME(bin_idx);
+	name_len = strlen(bin_name);
+
+	/* Get available size on file system */
+	if (statfs(BINARY_MNT_PATH, &fs_buf) != OK) {
+		bmdbg("Failed to stat fs, %s\n", BINARY_MNT_PATH);
+		return BINMGR_OPERATION_FAIL;
+	}
+	size = fs_buf.f_bavail * 1024;
```
Comment (Taejun-Kwon) [inline os/kernel/binary_manager/binary_manager_getinfo.c]: f_bavail means that free block.
So size is depend on block size, it is changeable value, not fixed as 1024.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 14]
```diff
@@ -74,42 +75,44 @@ typedef struct binary_header_s binary_header_t;
...
-		bmdbg("Failed to read %s: %d, errno %d\n", devname, ret, errno);
+		bmdbg("Failed to read %s: %d, errno %d\n", path, ret, errno);
 		goto errout_with_fd;
 	}
 
 	/* Read the binary header */
 	ret = read(fd, (FAR uint8_t *)header_data, sizeof(binary_header_t));
 	if (ret != sizeof(binary_header_t)) {
-		bmdbg("Failed to read %s: %d, errno %d\n", devname, ret, errno);
+		bmdbg("Failed to read %s: %d, errno %d\n", path, ret, errno);
```
Comment (Taejun-Kwon) [inline os/kernel/binary_manager/binary_manager_load.c]: same comment as I wrote above. just read once.
I/O always takes much of time than any other operation.

File: os/kernel/binary_manager/binary_manager_load.c [hunk 15]
```diff
@@ -137,117 +141,157 @@ static int binary_manager_read_header(int bin_idx, int part_idx, binary_header_t
...
+		binfmt_exit(binp);
+		binp = NULL;
+		BIN_LOADINFO(bin_idx) = NULL;
+	}
+	load_attr.binp = binp;
 #endif
+	bmvdbg("BIN[%d] %s %d %d\n", bin_idx, path, load_attr.bin_size, load_attr.offset);
+
+	retry_count = 0;
+	while (retry_count < BINMGR_LOADING_TRYCNT) {
```
Comment (Taejun-Kwon) [inline os/kernel/binary_manager/binary_manager_load.c]: why retry is necessary??any specific reason?

**Comment by seinfra** [comment]
Target : [4d74368a75380e27f61607fb138368421ec872da] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [4d74368a75380e27f61607fb138368421ec872da] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fe6cb400440ecb76434b8ab79faa89e68609e89c] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [fe6cb400440ecb76434b8ab79faa89e68609e89c] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [22a00b8eb7571d958cb956125e1923d413549a8b] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [22a00b8eb7571d958cb956125e1923d413549a8b] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [8109fdf0bdbdd131f07a2e9bdcbcaa8176738c61] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [8109fdf0bdbdd131f07a2e9bdcbcaa8176738c61] - Code Rule Check (C++) OK.
